website/src/client/components/Router.tsx

```diff
@@ -6,42 +7,60 @@ import App from './App';
 import EmbeddedApp from './EmbeddedApp';
 import NonExistent from './NonExistent';
 
-type Props = {
+type RouterProps = {
+  /** The server context if Snack is being server-side rendered */
+  ctx?: ServerContext;
   data: RouterData;
   queryParams: QueryParams;
   userAgent: string;
 };
 
-export default class Router extends React.Component<Props> {
-  private _renderRoute = (props: any) => {
-    const { data, ...rest } = this.props;
-    const isEmbedded = props.location.pathname.split('/')[1] === 'embedded';
+export default function Router(props: RouterProps) {
+  const render404 = React.useCallback(() => {
+    if (props.ctx) {
+      // If te router is server-side rendered, we need to set the right response code and headers.
+      // Without this, possible CDNs might cache the invalid path.
+      props.ctx.response.status = 404;
+      props.ctx.response.set(
+        'Cache-Control',
+        'private, no-cache, no-store, max-age=0, must-revalidate'
+      );
+    }
+
+    return <NonExistent />;
+  }, []);
+
+  const renderRoute = React.useCallback(
+    (routeProps: any) => {
+      const { data, ...rest } = props;
+      const isEmbedded = routeProps.location.pathname.split('/')[1] === 'embedded';
+
+      if (!data || data.type !== 'success') {
+        return render404();
+      }
 
-    if (data && data.type === 'success') {
       const appProps = {
-        ...props,
+        ...routeProps,
         ...rest,
-        query: this.props.queryParams,
+        query: props.queryParams,
         snack: data.snack,
         defaults: data.defaults,
       };
+
       return isEmbedded ? <EmbeddedApp {...appProps} /> : <App {...appProps} />;
-    } else {
-      return <NonExistent />;
-    }
-  };
-
-  render() {
-    return (
-      <Switch>
-        <Route exact path="/embedded/@:username/:projectName+" render={this._renderRoute} />
-        <Route exact path="/embedded/:id" render={this._renderRoute} />
-        <Route exact path="/embedded" render={this._renderRoute} />
-        <Route exact path="/@:username/:projectName+" render={this._renderRoute} />
-        <Route exact path="/:id" render={this._renderRoute} />
-        <Route exact path="/" render={this._renderRoute} />
-        <Route component={NonExistent} />
-      </Switch>
-    );
-  }
+    },
+    [render404]
+  );
```

Comment:
List out the props instead of using `...rest` and specify them as dependencies for the callback hook: data, queryParams, userAgent. This also makes sure ctx doesn't get passed into the components.

Reply:
The old `renderRoute` would always read from the latest props (`this.props`).

This new code is buggy if the Router component is re-rendered: since `render404` is the only Hook dependency, if it doesn't change, then we'll use the old, memoized value of `renderRoute` from a previous render, which references the old `props` variable.

To match the behavior of the code for this PR, there needs to be no dependency array: always redefine `renderRoute` so that it always references the latest `props` variable.

Reply:
Yep, I agree on the rerender part. But in SSR we don't rerender, and in the client, all the `rest` props from `props` are only used in `App` when actually initially mounting the component. Switching from embed to app (or reversed) isn't possible/irrelavant client-side, error pages (e.g. snack data has an error) only contain an absolute link that will trigger a reload (I believe), and the router props arent part of the rendering stuff.

That being said, I think it's fine to just create a new renderRoute function for a rerender. We don't actually do that in the Snack app itself, we just use these routes to start the App in different ways.

Reply:
Done, also the only item in `rest` turned out to be `userAgent`, so I updated that. Also moved it to a function instead of callback to avoid possible (but unlikely) `data` content changes.